src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1111,17 +1097,17 @@ private enum AssignmentKind
             }
             else if (useLegacyWarnings)
             {
+                if (isMaybeDefaultValue(valueType))
+                {
+                    // No W warning reported assigning or casting [MaybeNull]T value to T
+                    // because there is no syntax for declaring the target type as [MaybeNull]T.
+                    return;
+                }
                 ReportNonSafetyDiagnostic(location);
             }
-            else if (assignmentKind == AssignmentKind.ForEachIterationVariable && isMaybeDefaultValue(valueType))
-            {
-                // No warning reported assigning [MaybeNull]T value to foreach iteration variable
-                // because there is no syntax for declaring the variable as [MaybeNull]T.
-                return;
-            }
             else
             {
-                ReportDiagnostic(assignmentKind switch { AssignmentKind.Return => ErrorCode.WRN_NullReferenceReturn, AssignmentKind.ForEachIterationVariable => ErrorCode.WRN_NullReferenceIterationVariable, _ => ErrorCode.WRN_NullReferenceAssignment }, location);
+                ReportDiagnostic(assignmentKind switch { AssignmentKind.Return => ErrorCode.WRN_NullReferenceReturn, _ => ErrorCode.WRN_NullReferenceAssignment }, location);
```

Comment:
>switch { AssignmentKind.Return => ErrorCode.WRN_NullReferenceReturn, _ => ErrorCode.WRN_NullReferenceAssignment } [](start = 48, length = 113)

Consider simplifying to a ternary. I used a switch when there were 3 options because I felt it was clearer, but now that we have just 2 the ternary is simpler.